Update to latest ELK stack

mrdavidlaing opened the pull request:
The PR upgrades all the ELK components to:
1.  Java8
2.  Elasticsearch 1.4.4
3.  Logstash 1.5.0.rc2
4.  Kibana 4.0.1

mrdavidlaing committed:
Update to Kibana 4.0.1 and Elasticsearch 1.4.4

mrdavidlaing committed:
Update to logstash-1.5.0.rc2

mrdavidlaing committed:
Upgrade to OpenJDK 1.8.0_40

mrdavidlaing commented on packages/logstash/packaging:
```diff
@@ -6,5 +6,16 @@ set -u # report the usage of uninitialized variables
+# $BOSH_INSTALL_TARGET/logstash/bin/plugin install $BOSH_COMPILE_TARGET/logstash/logstash-filter-alter-0.1.5.gem
+
+# For now, we have to install from rubygems; meaning that the compilation VMs must have 'net access
+$BOSH_INSTALL_TARGET/logstash/bin/plugin install logstash-filter-alter
```
@dpb587 - Any ideas on what to do w.r.t which logstash 1.5.0 plugins to install?

What do you think to just installing them all?

dpb587 replied:
After the referenced bug is fixed. when installing .gem files, is an internet connection required? What about for more complex ones which may have a dependency on another third-party gem?

mrdavidlaing replied:
> when installing .gem files, is an internet connection required

In theory no.

However, 

> What about for more complex ones

I don't know :disappointed:

mrdavidlaing committed:
Include logstash-filter-translate and logstash-input-relp in log stash package

mrdavidlaing committed:
Revert to using Java7 for ingestor_syslog, to retain compatibility for SSLv3 used by nxlog v2.8.1248 (nxlog-ce-2.9.1347+ supports TLS 1.2, which is used by Java8)

mrdavidlaing committed:
Connect Kibana to first elasticsearch.host when it is a comma separated list

mrdavidlaing committed:
Bugfix: Use Java7 for ingestor_syslog

dpb587 commented:
You're brave upgrading both software and its runtime in a single step. Personally I would have been a bit more conservative and done java after we successfully deployed and confirmed the other upgrades. But, if you're confident everything is working, I'm good with that.

I didn't realize logstash 1.5 hadn't been release yet - are we going to wait to merge until 1.5 is stable and we've updated this branch with it?

mrdavidlaing commented:
It turns out that this upgrade is going to be breaking; so I'm inclined to bundle all the breaks into one release.

Specifically:
- Logstash 1.5 breaks some filters (specificaly those that muck about with ruby `Time`)
- Java8 disables SSLv3, which breaks nxlog < 2.9

mrdavidlaing commented:
> are we going to wait to merge until 1.5 is stable and we've updated this branch with it

Lets wait up to 2 weeks, and then make this call again.

dpb587 commented:
logsearch-shipper then also needs to be updated and redeployed to all the deployments using it - otherwise it sounds like encrypted syslog forwarding is going to break when the target logsearch is upgraded.

mrdavidlaing commented:
Yes - logsearch-shipper needs to be upgraded

mrdavidlaing committed:
Use Java8 for ingestor_syslog, but allow SSLv3 support to be enabled via logstash_ingestor.syslog_tls.jdk_tls_client_protocols setting
ie: logstash_ingestor:
	syslog_tls:
		jdk_tls_client_protocols:"SSLv3,TLSv1,TLSv1.1,TLSv1.2"

mrdavidlaing committed:
Merge branch 'develop' into update_to_latest_ELK

* develop:
  Remove CHANGELOG - reference tags/releases instead

mrdavidlaing commented:
@dpb587, You can now enable SSLv3 support (it's disabled by default) by adding this to your deploy manifest:

```
logstash_ingestor:
    syslog_tls:
        jdk_tls_client_protocols:"SSLv3,TLSv1,TLSv1.1,TLSv1.2"
```

`logsearch-shipper` should still be upgraded to nxlog 2.9, but it isn't required to deploy logsearch v19

dpb587 commented:
Great, thanks.

mrdavidlaing committed:
Pull logsearch/logsearch-filters-common into logsearch/logsearch-boshrelease/src/logsearch-filters-common and always include "transport" related filters (syslog, relp).

mrdavidlaing committed:
Ensure that <% in src/logsearch-filters-common/target/logstash-filters-default.conf is escaped so that it isn't lost when BOSH parses it as a template

mrdavidlaing commented:
@dpb587, I've pulled `logsearch/logsearch-filters-common` into `src/logsearch-filters-common` as we discussed.

After removing (see https://github.com/cityindex/logsearch-filters-internal/pull/55) the `syslog` filters from `cityindex/logsearch-filters-internal`, the resulting running `/var/vcap/jobs/log_parser/config/logstash.conf` is the same, as expected.

I'm fairly pleased with the outcome.  Any further thoughts?

mrdavidlaing committed:
Add kibana4.meta.logsearch.io proxy - prevent startup timeout

mrdavidlaing committed:
EXPERIMENTAL: Relax api proxy rules to enable Kibana4 to work

mrdavidlaing commented:
BUG:  It turns out that for SSLv3 to be supported the `/var/vcap/packages/java8/lib/security/java.security` file must be edited and `SSLv3` removed from the `jdk.tls.disabledAlgorithms=SSLv3` setting

dpb587 commented on jobs/kibana/spec:
```diff
@@ -11,9 +10,9 @@ templates:
-  kibana.elasticsearch:
-    description: "IP address and port of elasticsearch host to proxy requests for (eg, 127.0.0.1:9200)"
-    default: "http://127.0.0.1:9200"
+  kibana.api_host:
```
Can we keep this as elasticsearch? Since kibana doesn't need to know the difference on whether it's logsearch's api or elasticsearch's api, it seems better to stick with expected vendor conventions.

mrdavidlaing replied:
OK

dpb587 commented on jobs/kibana/templates/bin/kibana_ctl:
```diff
@@ -5,14 +5,22 @@ set -u # report the usage of uninitialized variables
+      curl -s <%= p('kibana.api_host') %> | grep -q 200 && break
+      n=$[$n+1]
+      sleep 5s
+    done
```
Can you explain to me why this whole block is necessary?

dpb587 replied:
Answer: kibana start up will fail if elasticsearch isn't available

dpb587 commented on jobs/kibana/templates/config/kibana.conf.erb:
```diff
@@ -5,35 +5,56 @@ port: <%= p('kibana.port') %>
+# Time in milliseconds to wait for responses from the back end or elasticsearch.
 # This must be > 0
-request_timeout: 60
+request_timeout: 300000
```
These shard and request timeouts are heavily data and cluster dependent. I think they should be configurable (and defaulted to kibana's built-in defaults, if they're not already).

mrdavidlaing replied:
OK

dpb587 commented on jobs/kibana/templates/bin/kibana_ctl:
```diff
@@ -5,14 +5,22 @@ set -u # report the usage of uninitialized variables
+    done
 
-    echo $$ > $PIDFILE
+    pid_guard $PIDFILE kibana
```
This is using exec - if `/var/vcap/packages/kibana/bin/kibana` is forking, the `exec` term should be removed; otherwise, if it's not forking and as the parent process `$PIDFILE` should remain `$$`.

dpb587 commented on jobs/log_parser/templates/bin/log_parser_ctl:
```diff
@@ -26,11 +26,10 @@ case $1 in
     <% if '' != p('logstash_parser.filters') %>
       cat ${JOB_DIR}/config/filters_override.conf >> ${JOB_DIR}/config/logstash.conf
-    <% else %>
-      cat ${JOB_DIR}/config/filters_default.conf >> ${JOB_DIR}/config/logstash.conf
```
Note: this means we're no longer shipping some default filters that the experimental user can play with which I think we may be mentioned in the docs somewhere (which I'm perfectly okay with and support, but not something we previously discussed).

mrdavidlaing replied:
Agreed.  In logsearch v19+, you get bundled filters which do "transport" related things, like parse syslog; but nothing further.

dpb587 commented on src/logsearch-filters-common/Rakefile:
```diff
@@ -0,0 +1,37 @@
+require 'erb'
```
This file has been unused since we have `bin` scripts. It should be deleted.

dpb587 replied:
Apparently I was wrong. Nevermind.

dpb587 commented on src/logsearch-filters-common/target/logstash-filters-default.conf.erb:
```diff
@@ -0,0 +1,128 @@
+if [@type] in ["syslog", "relp"] {
```
I'm not a huge fan of having two compiled files committed. On the one hand, this is a static file, so it's much more appropriate to be included in the logstash package, bypassing the ERB double-escaping rationale for this particular file. On the other hand, I do feel like we might have a need to make something here user-configurable (e.g. default timezone).

mrdavidlaing replied:
After agreeing that this should be moved into the logstash package, I actually think that your final argument `we might have a need to make something here user-configurable` is actually the clincher as to why it should remain in the job.  (which in turn requires commiting 2 copies of the compiled file; one which "normal", and one which is "escaped" so it can be passed through BOSH's template generating ERB process)

mrdavidlaing committed:
Rename kibana.api_host to kibana.elasticsearch to better match kibana.yml property naming

mrdavidlaing committed:
Add kibana.request_timeout and kibana.shard_timeout properties to Kibana job.

mrdavidlaing committed:
Ensure that Java security config doesn't prevent using the SSLv3 protocol when we explicitly want to allow it.

mrdavidlaing commented:
@dpb587, could you peer review these changes please

mrdavidlaing commented:
@dpb587 I've deployed this to meta.logsearch.io - https://github.com/cityindex/labs-operations/compare/2135b58017d5...9719435c4add - and it seems to be working ok.

mrdavidlaing commented:
@dpb587, Its also deployed to cityindex.logsearch.io - https://github.com/cityindex/labs-operations/compare/9719435c4add...cbfe776225d4

If you are happy after review, please just go ahead and merge this PR.
